src/pdm/workqueue/Worker.py

```
@@ -296,7 +296,13 @@ def run(self):
                     json.dump(data, self._current_process.stdin)
                     self._current_process.stdin.write('\n')
                     self._current_process.stdin.flush()
+                    # We have to close stdin to force the subprocess to handle the input
+                    # Otherwise it assumes there may be more data and hangs...
+                    self._current_process.stdin.close()
                     stderr_dispatcher = BufferingDispatcher(self._current_process.stderr)
                     StdOutDispatcher(self._current_process.stdout, token_map,
                                      stderr_dispatcher, self._upload)
                     asyncore.loop(timeout=2)
+                    if self._current_process.wait():
+                        self._logger.error("Job %s failed", job['id'])
+                        self._logger.info("Job stderr:\n%s", stderr_dispatcher.buffer)
```

I need to think about this. The problem with checking the subprocess returncode is that the subprocess may complete ok (returncode 0) but one of the individual elements may have failed. This is why @martynia returns `Code` in his dict. Is the idea here to be a catch for when the subprocess itself fails rather than a given element, I guess so right?

---

Yes, the idea is to catch the "big" errors like gfal2 not being installed, or the wrong path to the script... Things that cause the entire interpreter to bail out with code != 0.

---

ok